design/ceph/ceph-stretch-cluster.md

```diff
@@ -0,0 +1,224 @@
+supported by OSDs can be used.
+
+In the minimum configuration, two nodes in each data zone would be labeled, while one
+node in the arbiter zone is required.
+
+For example:
+
+```yaml
+topology.kubernetes.io/zone=a
+topology.kubernetes.io/zone=a
+topology.kubernetes.io/zone=b
+topology.kubernetes.io/zone=b
+topology.kubernetes.io/zone=arbiter
+```
+
+### Rook Cluster Design
+
+The core changes to rook are to associate the mons with the required zones.
+
+- Five mons are required. No other count of mons is supported for stretch clusters.
+- One mon will be assigned to the arbiter zone
+- Two mons will each be assigned to the data zones
+  - The two mons within each zone will have node antiaffinity
+
+The new configuration is found under the `stretchCluster` configuration where the
+three zones must be listed and the arbiter zone is identified.
+
+```yaml
+  mon:
+    count: 5
+    allowMultiplePerNode: false
+    stretchCluster:
+      zones:
+      # There must be exactly three zones in this list, with one of them being the arbiter
+      - name: arbiter
+        arbiter: true
+      - name: a
+      - name: b
+```
+
+The operator will track which mons belong to which zone. The zone assignment for each mon
+will be stored in the `rook-ceph-mon-endpoints` configmap in the same data structure where the host
+assignment is stored for mons with node affinity.
+
+For example, if the zones are called
+"arbiter", "zone1", and "zone2", the configmap would contain:
+
+```
+data:
+  data: c=10.96.86.248:6789,a=10.99.109.200:6789,b=10.98.18.147:6789
+  mapping: '{"node":{"a":"arbiter","b":"zone1","c":"zone2"}}'
+  maxMonId: "2"
+```
+
+### Mon Failover
+
+Mon failover will replace a mon in the same zone where the previous mon failed.
+
+### Mon Storage
+
+The mons can all be backed by a host path or a PVC, similar to any other Rook cluster.
+
+In this example, all the mons will be backed by the same PVC:
+
+```yaml
+  mon:
+    count: 5
+    allowMultiplePerNode: false
+    stretchCluster:
+      zones:
+      - name: arbiter
+        arbiter: true
+      - name: a
+      - name: b
+    volumeClaimTemplate:
+      spec:
+        storageClassName: gp2
+        resources:
+          requests:
+            storage: 10Gi
+```
+
+It is possible that the mons will need to be backed by a different type of storage in different zones.
+In that case, a volumeClaimTemplate specified under the zone will override the default storage setting for
+the mon.
+
+For example, the arbiter could specify a different backing store in this manner:
+
+```yaml
+  mon:
+    count: 5
+    allowMultiplePerNode: false
+    stretchCluster:
+      zones:
+      - name: arbiter
+        arbiter: true
+        volumeClaimTemplate:
+          spec:
+            storageClassName: alternative-storage
+              resources:
+                requests:
+                  storage: 10Gi
+      - name: a
+      - name: b
+    volumeClaimTemplate:
```

Comment:
wrong indent

Reply:
Actually it was intended. The default volumeClaimTemplate would be specified at this higher level, while the template can be overridden inside each zone.

Reply:
Right!